pydantic_ai_slim/pydantic_ai/_parts_manager.py

```diff
@@ -120,43 +163,60 @@ def handle_text_delta(
-                        self._vendor_id_to_part_index.pop(vendor_part_id)
-                        return None
-                    else:
-                        return self.handle_thinking_delta(
-                            vendor_part_id=vendor_part_id, content=content, provider_details=provider_details
-                        )
+                        self._handle_embedded_thinking_end(vendor_part_id)
+                        return
+                    yield from self._handle_embedded_thinking_content(vendor_part_id, content, provider_details)
+                    return
                 elif isinstance(existing_part, TextPart):
                     existing_text_part_and_index = existing_part, part_index
                 else:
                     raise UnexpectedModelBehavior(f'Cannot apply a text delta to {existing_part=}')
 
         if thinking_tags and content == thinking_tags[0]:
             # When we see a thinking start tag (which is a single token), we'll build a new thinking part instead
-            self._vendor_id_to_part_index.pop(vendor_part_id, None)
-            return self.handle_thinking_delta(
-                vendor_part_id=vendor_part_id, content='', provider_details=provider_details
-            )
+            yield from self._handle_embedded_thinking_start(vendor_part_id, provider_details)
+            return
 
         if existing_text_part_and_index is None:
             # This is a workaround for models that emit `<think>\n</think>\n\n` or an empty text part ahead of tool calls (e.g. Ollama + Qwen3),
             # which we don't want to end up treating as a final result when using `run_stream` with `str` a valid `output_type`.
             if ignore_leading_whitespace and (len(content) == 0 or content.isspace()):
-                return None
+                return
 
             # There is no existing text part that should be updated, so create a new one
-            new_part_index = len(self._parts)
             part = TextPart(content=content, id=id, provider_details=provider_details)
-            if vendor_part_id is not None:
-                self._vendor_id_to_part_index[vendor_part_id] = new_part_index
-            self._parts.append(part)
-            return PartStartEvent(index=new_part_index, part=part)
+            new_part_index = self._append_part(part, vendor_part_id)
+            yield PartStartEvent(index=new_part_index, part=part)
         else:
             # Update the existing TextPart with the new content delta
             existing_text_part, part_index = existing_text_part_and_index
             part_delta = TextPartDelta(content_delta=content, provider_details=provider_details)
             self._parts[part_index] = part_delta.apply(existing_text_part)
-            return PartDeltaEvent(index=part_index, delta=part_delta)
+            yield PartDeltaEvent(index=part_index, delta=part_delta)
+
+    def _update_raw_content(
+        self,
+        part: ThinkingPart,
```

Comment:
this one handles the case you described in https://github.com/pydantic/pydantic-ai/pull/3559#discussion_r2566313092, where during streaming a model may first generate raw CoT and **then** a summary. I wasn't able to reproduce it with openrouter, but I though, better to handle it now than in the future. So this can stay or go, lmk what you think.